src/openms/include/OpenMS/ANALYSIS/ID/AhoCorasickDA.h

```diff
@@ -0,0 +1,245 @@
+// ETH Zurich, and Freie Universitaet Berlin 2002-2020.
+//
+// This software is released under a three-clause BSD license:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Chris Bielow$
+// $Authors: Patricia Scheil$
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <OpenMS/CHEMISTRY/AASequence.h>
+#include <OpenMS/CONCEPT/Types.h>
+#include <vector>
+#include <unordered_map>
+
+
+namespace OpenMS
+{
+
+  class OPENMS_DLLAPI AhoCorasickDA
+  {
+
+  public:
+
+    /// constructor
+    explicit AhoCorasickDA(const std::vector<String>& seq) : sequences_(seq)
+    {
+      construct_();
+    }
+
+    /// Destructor
+    virtual ~AhoCorasickDA() = default;
+
+
+
+    //needed for the construction
+    //contains informations about used nodes, used base values and empty elements
+    struct BuildInformation
+    {
+      UInt32 next        :31 ;    //next empty node in DA
+      UInt32 used_flag   :1  ;    //node is used
+      UInt32 prev        :31 ;    //previous empty node in the DA
+      UInt32 base_flag   :1  ;    //this value is used as base
+      UInt32 check           ;    //parent node TODO not necessary
+    };
+
+    struct BC
+    {
+      UInt32 base       :22 ;    // base value of DA
+      UInt32 lcheck     :8  ;    // arc label from parent node
+      UInt32 leaf_flag  :1  ;    // node is leaf
```

Comment:
one could play around with the order of the members here, and see what performs the best (since extraction of a single bit is not cheap) and extraction of lcheck could be cheaper if moved to the back...
(keep in mind for later maybe)